spacemacs-layouts: Fix adding buffers to new persp

Miciah opened the pull request:
Delete the hook that https://github.com/syl20bnr/spacemacs/pull/11642 added to `persp-created-functions`, and change `spacemacs/ivy-spacemacs-layouts`, `spacemacs/helm-perspectives`, `spacemacs/helm-persp-switch-project`, and `spacemacs/ivy-persp-switch-project` to achieve the intended goal of adding the desired buffers after creating a new perspective.

Change `spacemacs/helm-persp-switch-project` and `spacemacs/ivy-persp-switch-project` as follows: If the user selects a project but then quits without selecting a file or buffer, the new perspective is now immediately killed.  Otherwise, if the perspective did not already exist, any buffers that belong to the selected project are added to the perspective.

Add the following actions to `spacemacs/ivy-spacemacs-layouts` and `spacemacs/helm-perspectives`:

* Create a new perspective with the Spacemacs home buffer (default action).
* Create a new perspective with the buffers that belong to the current buffer's project.
* Create a new perspective with the buffers that belong to the current perspective (i.e., make a copy of the current perspective).

This commit resolves the problem reported in https://github.com/syl20bnr/spacemacs/commit/9fcf8c898dd1069ef362f1ede320e85065358ac6#commitcomment-33343455.

This commit also fixes a problem with the default action for `spacemacs/ivy-spacemacs-layouts` and `spacemacs/helm-perspectives`.  According to https://github.com/syl20bnr/spacemacs/pull/4077, the default action is supposed to display the home buffer if the action creates a new perspective, which was determined by checking whether the perspective was missing from the list of perspectives _before_ switching.  However, commit eb7ca651fe72241a5784c289543f73b6c273faf4 changed this logic so it was checking whether the perspective was missing from the list of perspectives _after_ switching.  This commit restores the correct logic.

* `CHANGELOG.develop`: Update.
* `layers/+completion/ivy/funcs.el` (`spacemacs/ivy-spacemacs-layouts`): Add "Copy Current Layout" and "Create Project Layout" actions to the docstring.  Use `spacemacs//create-persp-with-home-buffer` for the default action.
* `layers/+completion/ivy/packages.el` (`ivy/post-init-persp-mode`): Add actions for `spacemacs//create-persp-with-current-project-buffers` and `persp-copy`.
* `layers/+spacemacs/spacemacs-layouts/funcs.el` (`spacemacs||switch-layout`): New macro.  Switch to the named perspective, and initialize it using the provided forms if the perspective is new.
(`spacemacs//create-persp-with-current-project-buffers`): New function.  Create a new perspective with the current project's buffers.
(`spacemacs||switch-project-persp`): New macro.  Switch to the named perspective, and evaluate the provided forms with `projectile-after-switch-project-hook` bound with a hook that adds the current project's buffers to the perspective.  If the user quits during the evaluation of the forms, kill the perspective.
(`spacemacs//create-persp-with-home-buffer`): New function.  Switch to the named perspective, and go to the Spacemacs home buffer if the perspective is new.
(`spacemacs/helm-perspectives`): Use `spacemacs//create-persp-with-home-buffer` for the default action.  Add actions for `spacemacs//create-persp-with-current-project-buffers` and `persp-copy`.
(`spacemacs//helm-persp-switch-project-action`): New function.  Switch to the named perspective and call `projectile-switch-project-by-name`, using the new `spacemacs||switch-project-persp` macro.  Bind `helm-quit-hook` with a hook that kills the new perspective if the user quits `projectile-switch-project-by-name`.
(`spacemacs/helm-persp-switch-project`): Use `spacemacs//helm-persp-switch-project-action`.
(`spacemacs//ivy-persp-switch-project-action`): New function.  Switch to the named perspective and call `counsel-projectile-switch-project-action` with a hook to add the project's buffers to the new perspective.
(`spacemacs/ivy-persp-switch-project`): Delete advice for `counsel-projectile-switch-project-action`.  Use `spacemacs//ivy-persp-switch-project-action` instead.
(`spacemacs//add-project-buffers-to-persp`): Deleted.
* `layers/+spacemacs/spacemacs-layouts/packages.el`
(`spacemacs-layouts/init-persp-mode`): Don't add a hook to `persp-created-functions`.

duianto commented:
### Before
The same behavior was observed in both Helm and Ivy.

buffers (`SPC b b`) visible on startup: `spacemacs, scratch, messages`
open a new project layout: `SPC p l`
project: `dir-a`
file: `file-a`
buffers visible: `file-a, spacemacs, scratch`

open a new project layout `SPC p l`:
project: `dir-b`
file: `file-b`
buffers visible: `file-b, file-a`


### After
The same behavior was observed in both Helm and Ivy.

buffers visible on startup: `spacemacs, scratch, messages`
open a new project layout: `SPC p l`
project: `dir-a`
file: `file-a`
buffers visible: `file-a`

open a new project layout: `SPC p l`
project: `dir-b`
file: `file-b`
buffers: `file-b`

It seems to be working as expected 👍 

### Issue

#### Ivy

When canceling the `[dir-a] Load buffer or file:` prompt:
- persp switch project `SPC p l`
- choose a project `RET`
- cancel the project file choice `C-g`

Then all three buffers from the default layout are added to the new project layout:
`spacemacs, scratch and messages`

But when a project file `file-a` is opened, then the three initial buffers are removed and only the project file is shown in the buffer list.

It seems like all buffers from the default layout is added to any new project layout when it's created.

Because after the steps above if a new project layout is created and canceled before opening a buffer file, then it has four buffers: `file-a, spacemacs, scratch and messages`

The same thing as in layout `dir-a` is observed in the new layout `dir-b`, if a project file (`file-b`) is opened, then `SPC b b` only shows `file-b`.

In **Helm** no buffers are shown in the new layout after canceling before choosing a project file.

Maybe ivy should behave the same as Helm.


#### Before this PR, in both Helm and Ivy

When canceling `SPC p l` before opening a project file.

Then the `spacemacs and scratch` buffers were added to the new project layout.

And opening a project file just added it to the buffer list for that layout, and it still showed the default buffers.


### Question 1

This occurs in both Helm and Ivy.

Layout created, even if the open project file buffer/prompt is canceled.

If the helm buffer: `Helm Projectile`
or the ivy prompt: `[<project name>] Load buffer or file:`

(They appear after choosing a project from the: `SPC p l` buffer/prompt.)

Are canceled with `C-g` before a project file is opened.

Then the project layout still gets created.

The layout probably gets created as soon as one choses a project in the `SPC p l` HELM Projectile Layouts buffer.

Should the assumption be that the user also wanted to cancel the layout creation since they canceled before opening a project file?

### Question 2

I don't know how to access the new helm actions:
There are no entries for: `Create new perspective ...`
When pressing: `C-z` in either:
- HELM Projectile Layouts buffer (`SPC p l`)
- HELM Projectile buffer (after pressing `RET` on a project)

And I'm not sure where one is supposed to use the ivy keys `F2` and `F3` keys.

At the `SPC p l` prompts: Switch to Project Perspective
Or after choosing a project: `2    [dir-a] Load buffer or file:`
Pressing:
`F2` opens the which key popup for 2C-command, with the keys `2` -> `2C-two-columns` etc.
`F3` shows: `Defining kbd macro...`

I'm might be misunderstanding how they are supposed to work.

### System info
The before steps were tested in Ubuntu 18.04 with Emacs 27.0.50. (virtualbox guest, with win10 as host)
The after steps in Windows 10 with Emacs 26.1.

On the latest develop branch 89aae06f86c6453d7a62fc271710af4e14749968 with updated packages.

Miciah commented:
Thank you for testing and for your comments!

> [In Ivy,] if a new project layout is created and canceled before opening a buffer file, then it has four buffers: `file-a, spacemacs, scratch and messages`

In this case, the perspective (layout) really has no buffers (`(persp-buffer-list)` evaluates to nil), and apparently this causes `ivy-switch-buffer` to display all buffers without restricting to those in the current perspective.  You can see the same behavior if you create a perspective with no buffers by evaluating the following code and then pressing <kbd>SPC bb</kbd>:

```emacs-lisp
(let ((persp-reset-windows-on-nil-window-conf t)) (persp-switch "foo"))
```

> In **Helm** no buffers are shown in the new layout after canceling before choosing a project file.
> 
> Maybe ivy should behave the same as Helm.

> Should the assumption be that the user also wanted to cancel the layout creation since they canceled before opening a project file?

That seems reasonable. I have updated this PR to delete the perspective if the user quits the file/buffer selection.

> I don't know how to access the new helm actions:
> There are no entries for: `Create new perspective ...`
> When pressing: `C-z` in either:
> 
> * HELM Projectile Layouts buffer (`SPC p l`)
> * HELM Projectile buffer (after pressing `RET` on a project)

These actions are under <kbd>SPC l l</kbd>.  For example, <kbd>SPC l l</kbd>, enter a name that does not match an existing perspective, and press <kbd>F2</kbd> to create the perspective with any buffers that belong to the current buffer's project.

I have updated the PR with the change mentioned above to kill a new perspective if the user quits the file/buffer selection, as well as some further refactoring.

duianto commented:
### Actions

>These actions are under SPC l l. For example, SPC l l, enter a name that does not match an existing perspective, and press F2 to create the perspective with any buffers that belong to the current buffer's project.

I hadn't used actions from `SPC l l` before, but now I know about them, thanks.

![emacs_2019-05-12_12-03-03](https://user-images.githubusercontent.com/13420573/57582393-985aa480-74c4-11e9-9f52-467f7c7317fc.png)


### Issue 1, Helm only

Canceling at the project file selection
`SPC p l` select a project `RET` `C-g`
shows:
>persp-contain-buffer-p: Wrong type argument: perspective, :nil

**Ivy** seems to be working as expected. There's no error message.

### Issue 2, Helm and Ivy

The third action:
"Create new perspective with buffers from current perspective"

doesn't copy the buffers to the new perspective, the new perspective doesn't have any buffers.

#### Possible cause
It might only be working from the `Default` layout, because when a new layout is created from the `Default` layout, then a prompt appears:
>What buffers to copy?  (all, displayed, free and displayed, [F] free, choose, none, ?):

When calling the third action from any other layout, then there's no prompt and the new layout doesn't have any buffers.


### Ivy keys

The commit message mentions that the actions are:
>available via a key map in Ivy

And `F2` and `F3` works from `SPC l l`

But is the Ivy key map visible somewhere in the UI?

Or can the actions be added to the `M-o` actions list?

Currently it has entries for:
[c] Close layout(s)
[k] Kill layout(s)


### Typo

In the docstring for the function: `spacemacs//create-persp-with-current-project-buffers`

>   "Switch to perspective and add project buffers.
If perspective NAME does not already exist, create it and add any
buffers that belong the current buffer's project.  If the
perspective already exists, just switch to it."


Missing "to" on the third line, between:
"that belong"
and
"the current"


### Changelog.develop

You'll know best how to summarize this PRs changes.

Miciah commented:
> ### Issue 1, Helm only
> Canceling at the project file selection
> `SPC p l` select a project `RET` `C-g`
> shows:
> 
> > persp-contain-buffer-p: Wrong type argument: perspective, :nil

Fixed.  Thanks! (For some reason, this did not happen with my `.spacemacs` configuration, but I was able to reproduce it using the `.spacemacs` template.)

> ### Issue 2, Helm and Ivy
> The third action:
> "Create new perspective with buffers from current perspective"
> 
> doesn't copy the buffers to the new perspective, the new perspective doesn't have any buffers.

You probably need two upstream fixes to persp-mode that were merged yesterday.  I cannot reproduce this issue using persp-mode-20190511.1402.
 
> #### Possible cause
> It might only be working from the `Default` layout, because when a new layout is created from the `Default` layout, then a prompt appears:
> 
> > What buffers to copy?  (all, displayed, free and displayed, [F] free, choose, none, ?):

It looks like the upstream change https://github.com/Bad-ptr/persp-mode.el/commit/e330e6240bbb82589077f30472b05b95d1ff430d?w=1#diff-ad587a8d68b99fcd1d26681ee77cccffL2589 changes this behavior so you should no longer see this prompt.

> When calling the third action from any other layout, then there's no prompt and the new layout doesn't have any buffers.

Upstream change https://github.com/Bad-ptr/persp-mode.el/pull/103 should fix this problem.  (Sorry, I should have mentioned that change, which was pending when I made this PR.)

> ### Ivy keys
> The commit message mentions that the actions are:
> 
> > available via a key map in Ivy
> 
> And `F2` and `F3` works from `SPC l l`
> 
> But is the Ivy key map visible somewhere in the UI?
> 
> Or can the actions be added to the `M-o` actions list?
> 
> Currently it has entries for:
> [c] Close layout(s)
> [k] Kill layout(s)

Yes, thank you for pointing out <kbd>M-o</kbd>! I usually use Helm and was ignorant of this Ivy feature.  I have updated this PR to delete the key map and instead add the actions to the existing actions list.

> ### Typo
> In the docstring for the function: `spacemacs//create-persp-with-current-project-buffers`
> 
> > "Switch to perspective and add project buffers.
> > If perspective NAME does not already exist, create it and add any
> > buffers that belong the current buffer's project.  If the
> > perspective already exists, just switch to it."
> 
> Missing "to" on the third line, between:
> "that belong"
> and
> "the current"

Thanks! Fixed.

> ### Changelog.develop
> You'll know best how to summarize this PRs changes.

Updated.  Thank you very much for your time and attention to detail!  I'm sure the code still has room for improvement, but your comments have greatly improved the quality of my initial submission.

duianto commented:
Everything seems to be working as expected, thanks 👍 

There's always room for improvements, but this is how far we can see at the moment.

The `changelog.develop` just needs to be updated, then this is ready for cherry-picking.

Miciah commented:
> The `changelog.develop` just needs to be updated, then this is ready for cherry-picking.

Done.  Thanks!

duianto commented:
Thank you for contributing to Spacemacs!
The PR has been cherry-picked into the `develop` branch, you can safely delete your branch.

ynilu commented:
@Miciah, thank you for the fix and improvement. It's now working properly, so I can remove my patch.
@duianto, thank you for the comprehensive test.
